core/services/log/pool.go

```diff
@@ -1,50 +1,111 @@
 package log
 
 import (
-	"fmt"
-
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
-	"github.com/smartcontractkit/chainlink/core/logger"
 	"github.com/smartcontractkit/chainlink/core/store/models"
+	heaps "github.com/theodesp/go-heaps"
+	pairingHeap "github.com/theodesp/go-heaps/pairing"
 )
 
 type (
 	logPool struct {
-		allLogs []types.Log
+		hashesByBlockNumbers map[uint64][]common.Hash
+		logsByBlockHash      map[common.Hash][]types.Log
+		heap                 *pairingHeap.PairHeap
 	}
 )
 
 func newLogPool() *logPool {
-	return &logPool{}
+	return &logPool{
+		hashesByBlockNumbers: make(map[uint64][]common.Hash),
+		logsByBlockHash:      make(map[common.Hash][]types.Log),
+		heap:                 pairingHeap.New(),
+	}
 }
 
 func (pool *logPool) addLog(log types.Log) {
-	pool.allLogs = append(pool.allLogs, log)
+	pool.hashesByBlockNumbers[log.BlockNumber] = append(pool.hashesByBlockNumbers[log.BlockNumber], log.BlockHash)
+	pool.logsByBlockHash[log.BlockHash] = append(pool.logsByBlockHash[log.BlockHash], log)
+	pool.heap.Insert(Uint64(log.BlockNumber))
 }
 
-func (pool *logPool) getLogsToSend(head models.Head, highestNumConfirmations uint64, finalityDepth uint64) []types.Log {
+func (pool *logPool) getLogsToSend(head models.Head, highestNumConfirmations uint64, finalityDepth uint64) ([]types.Log, int64, int64, int64) {
 	latestBlockNum := uint64(head.Number)
-	logsToReturn := pool.allLogs
-	var logsToKeep []types.Log
+	logsToReturn := make([]types.Log, 0)
 
 	keptLogsDepth := finalityDepth
 	if highestNumConfirmations > keptLogsDepth {
 		keptLogsDepth = highestNumConfirmations
 	}
-	// deleting old logs that will never be sent for any listener anymore
-	if len(pool.allLogs) > 0 {
-		for _, log := range pool.allLogs {
-			if int64(log.BlockNumber) >= int64(latestBlockNum)-int64(keptLogsDepth) {
-				logsToKeep = append(logsToKeep, log)
-			}
+
+	keptDepth := int64(latestBlockNum) - int64(keptLogsDepth)
```

Comment:
seems a little weird to compute this inside this func when its only used outside - can just compute it in deleteOldLogs

Reply:
good point - moving this out simplified it quite a bit